Userland/DevTools/HackStudio/ProjectTemplate.cpp

```diff
@@ -0,0 +1,279 @@
+#include <LibCore/DirIterator.h>
+#include <LibCore/File.h>
+#include <assert.h>
+#include <fcntl.h>
+#include <spawn.h>
+#include <sys/stat.h>
+#include <sys/wait.h>
+#include <unistd.h>
+
+// FIXME: shameless copy+paste from Userland/cp. We should have system-wide file management functions.
+// Issue #5209
+bool copy_file_or_directory(String, String, bool, bool);
+bool copy_file(String, String, const struct stat&, int);
+bool copy_directory(String, String, bool);
+
+namespace HackStudio {
+
+ProjectTemplate::ProjectTemplate(const String& id, const String& name, const String& description, const GUI::Icon& icon, int priority)
+    : m_id(id)
+    , m_name(name)
+    , m_description(description)
+    , m_icon(icon)
+    , m_priority(priority)
+{
+}
+
+RefPtr<ProjectTemplate> ProjectTemplate::load_from_manifest(const String& manifest_path)
+{
+    auto config = Core::ConfigFile::open(manifest_path);
+
+    if (!config->has_group("HackStudioTemplate")
+        || !config->has_key("HackStudioTemplate", "Name")
+        || !config->has_key("HackStudioTemplate", "Description")
+        || !config->has_key("HackStudioTemplate", "IconName32x"))
+        return {};
+
+    auto id = LexicalPath(manifest_path).title();
+    auto name = config->read_entry("HackStudioTemplate", "Name");
+    auto description = config->read_entry("HackStudioTemplate", "Description");
+    int priority = config->read_num_entry("HackStudioTemplate", "Priority", 0);
+
+    // Attempt to read in the template icons
+    // Fallback to a generic executable icon if one isn't found
+    auto icon = GUI::Icon::default_icon("filetype-executable");
+
+    auto bitmap_path_32 = String::formatted("/res/icons/hackstudio/templates-32x32/{}.png", config->read_entry("HackStudioTemplate", "IconName32x"));
+
+    if (Core::File::exists(bitmap_path_32)) {
+        auto bitmap32 = Gfx::Bitmap::load_from_file(bitmap_path_32);
+        icon = GUI::Icon(move(bitmap32));
+    }
+
+    return adopt(*new ProjectTemplate(id, name, description, icon, priority));
+}
+
+bool ProjectTemplate::create_project(const String& name, const String& path)
+{
+    // Check if a file or directory already exists at the project path
+    if (Core::File::exists(path))
+        return false;
+
+    dbgln("Creating project at path '{}' with name '{}'", path, name);
+
+    // Verify that the template content directory exists. If it does, copy it's contents.
+    // Otherwise, create an empty directory at the project path.
+    if (Core::File::exists(content_path()) && Core::File::is_directory(content_path())) {
+        copy_directory(content_path(), path, false);
+    } else {
+        dbgln("No template content directory found for '{}', creating an empty directory for the project.", m_id);
+        int rc;
+        if ((rc = mkdir(path.characters(), 0777)) < 0) {
+            warnln("Failed to mkdir empty project directory, error: {}, rc: {}", strerror(errno), rc);
+            return false;
+        }
```

Comment:
Could `File::ensure_parent_directories()` work here ? Maybe if you absolutely want to create it as `777` instead of `755`, but I don't really see the reason to do that specifically in the rest of the code (and `777` looks wrong for a subdirectory of home).

Reply:
It seems `ensure_parent_directories` wouldn't work in this situation, it appears to only creates the parent directories in a path, but not the trailing directory itself. `755` is definitely the more correct permission, I'll make sure to change it :)